CATS support

scd31 opened the pull request:
Tons to do still, but I'm working on adding [CATS](https://cats.radio) support. I only have a DFM-17 so I'm starting with that.

I'm using the FIFO for packet synthesis. This has numerous advantages as I outlined here: https://github.com/mikaelnousiainen/RS41ng/discussions/85 . Most notably, it means the accuracy of the mcu clock is no longer important. I intend to move all other modes to do this, assuming there are no objections. It should make APRS much more robust, as it won't need to have magic sleep values anymore.

Again, very WIP! Todo still:

- [x] Actually generate CATS packets (right now I generate them ahead of time and just hard-code them in the source code)
- [x] Remove all my changes to config.h that I just did for debugging

scd31 commented on src/hal/delay.h:
```diff
@@ -1,6 +1,8 @@
 #ifndef __DELAY_H
 #define __DELAY_H
 
+#include <stm32f10x.h>
```
I needed this to get the project to compile. I am not really sure why (or more specifically, I don't understand why it works without this!)

mikaelnousiainen replied:
I don't see how the STM32 header would be needed by this header file? Maybe it's missing somewhere else!

Though this header does use the stdint types, so `stdint.h` should be there!

So I'd change this.

scd31 commented on src/drivers/si4063/si4063.c:
```diff
@@ -218,10 +223,88 @@ void si4063_disable_tx()
+// If less than len, you will need to keep calling this
+uint16_t si4063_refill_buffer(uint8_t *data, int len)
+{
+    uint8_t response[2];
+
+    // Check how many bytes we have free
+    si4063_send_command(SI4063_COMMAND_FIFO_INFO, 0, NULL);
+    si4063_read_response(sizeof(response), response);
+
+    uint8_t free_space = response[1];
```
Very odd - this throws a warning that `response[1]` may not be initialized. It goes away if I change line 263 to `uint8_t response[3];`

scd31 replied:
Fixed by adding a `... = {0, 0};`. Seems weird this is needed when it's not needed for `response[3]`

mikaelnousiainen commented on src/codecs/cats/cats.c:
```diff
@@ -0,0 +1,34 @@
+
+cats_packet cats_create(uint8_t *data)
+{
+    cats_packet c = {.data = data, .len = 0};
```
What is `len` used for?

mikaelnousiainen replied:
Ok, yes, it's used later on, I just didn't immediately see it. I'm just wondering if there could be ways to make sure the packet will never grow outside of boundaries of the allocated `data` length.

scd31 replied:
What do we do for the APRS type?

The only way this would be able to grow too long is from the user specifying a very long message. We could check for this and throw a compile-time error, or we could do a bounds check at run-time and skip the CATS transmission if we'd overrun the buffer

mikaelnousiainen replied:
Yeah, let's leave this maybe as a future enhancement. We'll have to think about a good approach.

mikaelnousiainen commented on src/codecs/cats/crc.c:
```diff
@@ -0,0 +1,32 @@
+size_t cats_append_crc(uint8_t *data, size_t len)
+{
+    uint16_t crc = calc_crc(data, len);
+    data[len++] = crc;
```
With this kind of functions we must be careful to stay inside boundaries of allocated memory regions.

mikaelnousiainen commented on src/codecs/cats/ldpc.c:
```diff
@@ -0,0 +1,91 @@
+
+cats_ldpc_code_t *cats_ldpc_pick_code(int len)
+{
+    if(128 <= len)
```
Small thing, but let's try to stick with the formatting conventions:

- Spaces after keywords
- Always use braces `{}` and the way to use them (outside function definitions) is:

```
if (condition) {
  // TODO
} else {
  // TODO
}

mikaelnousiainen replied:
Also, I think the conditions read better when the variable is put first, e.g. len >= 128...

mikaelnousiainen commented on src/codecs/cats/ldpc.c:
```diff
@@ -0,0 +1,91 @@
+size_t cats_ldpc_encode_chunk(uint8_t *data, cats_ldpc_code_t *code, uint8_t *parity_out)
+{
+    // Code parameters
+    int k = code->k;
```
Not being familiar with the LDPC algorithm, would `k`, `r` and `b` be somehow self-explanatory for one who knows about them? I've tried to write code in this project so that it should be _mostly_ possible to read the code on its own -- so that the code documents itself. Could we possibly use some more descriptive variable names anyway (also in the struct)?

Also links to the algorithm documentation (or CATS documentation) would be great!

mikaelnousiainen commented on src/codecs/cats/ldpc.c:
```diff
@@ -0,0 +1,91 @@
+    int r = code->n - code->k;
+    int b = code->circulant_size;
+    const uint64_t* gc = code->matrix;
+    int rowLen = r/64;
```
Let's use `snake_case` naming

mikaelnousiainen commented on src/codecs/cats/ldpc.c:
```diff
@@ -0,0 +1,91 @@
+    // This means we can only encode up to 511 bytes
+    assert(len < 512);
+    
+    uint8_t parity[128];
```
Is the 128 byte array length based on some known limit? Could the array be smaller? There's not a lot of room for large arrays in stack...

scd31 replied:
See section 5.2 here: https://gitlab.scd31.com/cats/cats-standard/-/blob/master/standard.pdf?ref_type=heads

So we need a 128 byte array to support CATS packets up to 511 byte array. We could lower this to 32 bytes but we could only support up to 127 byte CATS packets then, due to how packets are broken down into chunks for LDPC encoding.

I'm not opposed to doing this if you think it's necessary. But I don't think this stack array should cause issues with other modes since it would only be "allocated" when CATS is doing its processing

mikaelnousiainen replied:
I'll add a comment about the size of the parity array!

mikaelnousiainen commented on src/codecs/cats/ldpc_matrices.h:
```diff
@@ -0,0 +1,26 @@
+#define __CATS_LDPC_MATRICES_H
+
+typedef struct cats_ldpc_code {
+    // Code length (data+parity)
```
Ok, now I see these comments. Could we just name the variables accordingly, so no comments would be necessary?

mikaelnousiainen commented on src/codecs/cats/whisker.c:
```diff
@@ -0,0 +1,125 @@
+#include "whisker.h"
+
+#ifdef RS41
+#define HARDWARE_ID 0x7d00
```
Are these some pre-allocated IDs for whisker data? This could deserve a comment and maybe even a link to CATS documentation where these are defined!

mikaelnousiainen commented on src/codecs/cats/whisker.c:
```diff
@@ -0,0 +1,125 @@
+#define HARDWARE_ID 0x7d01
+#endif
+
+#define SOFTWARE_ID 0x00
```
Again, maybe a comment would help here a bit (some CATS-specific ID?)

mikaelnousiainen commented on src/codecs/cats/whisker.c:
```diff
@@ -0,0 +1,125 @@
+
+void cats_append_identification_whisker(cats_packet *packet, char *callsign, uint8_t ssid, uint16_t icon)
+{
+    packet->data[packet->len++] = 0x00; // type = identidication
```
Typo in `identidication`

mikaelnousiainen replied:
Also, there could be (#define) constants for the types (and other "magic" numbers)

mikaelnousiainen commented on src/codecs/cats/whisker.c:
```diff
@@ -0,0 +1,125 @@
+    // bitmask
+    d[(*l)++] = 0;
+    d[(*l)++] = (CATS_IS_BALLOON << 2) | (has_altitude << 1);
+    d[(*l)++] = 0b11110011;
```
Could you add a comment for the bitmask or -- even better -- if each bit corresponds to some particular field being present, add them as #define macros and simply OR them here?

mikaelnousiainen commented on src/codecs/fifo/fifo.c:
```diff
@@ -0,0 +1,46 @@
+    uint16_t len;
+} fifo_encoder;
+
+void fifo_encoder_new(fsk_encoder *encoder)
```
This "encoder" could probably be named some other way as it's simply a "container" for data in this case, having nothing specific to any encoding (or use of FIFO in the radio chip).

Maybe it could be just called `raw_encoder` ?

scd31 replied:
Raw encoder makes sense to me

mikaelnousiainen commented on src/config.h:
```diff
@@ -335,6 +338,25 @@
+ * CATS mode settings
+ */
+#define CATS_CALLSIGN CALLSIGN
+#define CATS_SSID 29 // 0 - 255
```
Please add comments with links to the CATS documentation that specified SSIDs and icons. It'll help users a lot! :)

mikaelnousiainen commented on src/radio.c:
```diff
@@ -954,7 +981,7 @@ void radio_handle_main_loop()
                 current_telemetry_data.gps.latitude_degrees_1000000 / 10,
                 current_telemetry_data.gps.longitude_degrees_1000000 / 10,
-                (current_telemetry_data.gps.altitude_mm / 1000) * 3280 / 1000);
+                (current_telemetry_data.gps.altitude_mm / 1000));
```
:+1:

mikaelnousiainen reviewed (commented):
:+1: Looks great, thanks a lot for the contribution and all the (quite advanced) work on Si4063 FIFO mode + CATS encoding!

I've left a bunch of comments, but they're more on the "stylistic" side. I think more or less all code is where it should be.

The only thing which is a bit out of place is the "fifo encoder", where we now need to expose the data buffer... it's ok for now, but clearly the "fsk_encoder" abstraction doesn't fit this purpose.

mikaelnousiainen commented on src/radio_payload_cats.c:
```diff
@@ -0,0 +1,51 @@
+        *(cur++) = (CATS_SYNC_WORD >> (i * 8));
+    }
+
+    uint8_t *data = malloc(length);
```
Do we have any idea on what the maximum packet size is? It's easy to run out of memory here.

mikaelnousiainen commented on src/codecs/fsk/fsk.h:
```diff
@@ -38,6 +38,18 @@ typedef struct _fsk_encoder_api {
+     * @param encoder
+     * @return Pointer to packet data
+     */
+    uint8_t *(*get_data)(fsk_encoder *encoder);
```
Exposing the data here isn't really the right thing to do for this API struct, but since this "fsk encoder" is now the only encoder abstraction we have, let's go with this. :)

mikaelnousiainen commented on src/config.h:
```diff
@@ -335,6 +338,25 @@
 #define HORUS_V2_TIME_SYNC_OFFSET_SECONDS 0
 
+/**
+ * CATS mode settings
```
Please add a link to the CATS main website here

mikaelnousiainen commented:
@scd31 Can you also add some notes about the new mode in the README file (in relevant places regarding which hardware supports it)? Links to the CATS website and related documentation are welcome too (you can just follow how Horus is documented).

scd31 commented:
Thanks for the thorough review! I believe I've addressed all feedback. I also confirmed it compiles on the RS41.

Tomorrow I will do a thorough test and confirm that it works with multiple modes enabled for more than 30 minutes.

mikaelnousiainen commented on src/codecs/cats/ldpc.c:
```diff
@@ -0,0 +1,99 @@
+    // This means we can only encode up to 511 bytes
+    assert(len < 512);
+    
+    uint8_t parity[128];
```
```suggestion
    // A 128 byte array is needed to support CATS packets up to 511 bytes.
    uint8_t parity[128];
```

mikaelnousiainen commented:
> Thanks for the thorough review! I believe I've addressed all feedback. I also confirmed it compiles on the RS41.
> 
> Tomorrow I will do a thorough test and confirm that it works with multiple modes enabled for more than 30 minutes.

Thanks! Looks good now. Please tell me the results of your tests and we can then merge the PR!

scd31 commented:
Fixed a small bug I found. Other than that, I've had this running for about 4 hours with the following config:

```
#define RADIO_SI4063_TX_CW true
#define RADIO_SI4063_TX_CW_COUNT 1
#define RADIO_SI4063_TX_PIP true
#define RADIO_SI4063_TX_PIP_COUNT 6
#define RADIO_SI4063_TX_APRS true
#define RADIO_SI4063_TX_APRS_COUNT 2
#define RADIO_SI4063_TX_HORUS_V1 false
#define RADIO_SI4063_TX_HORUS_V1_COUNT 1
#define RADIO_SI4063_TX_HORUS_V2 true
#define RADIO_SI4063_TX_HORUS_V2_COUNT 6
#define RADIO_SI4063_TX_CATS true
#define RADIO_SI4063_TX_CATS_COUNT 5
```

Appears to be rock solid and is still decodeable!

mikaelnousiainen commented:
@scd31 Thanks a lot for testing (and the final bug fixes). I'll go ahead and merge this, as the additions should not break any existing features.
